### PR TITLE
feat: add CLI command to manage users

### LINKED ISSUE #566: CLI command to manage users
Hello. I wrote a command to manage users from CLI : create, activate, deactivate, change name, change email, set password

Here my last commit :
https://github.com/robertogerola/shield/commit/ab09040c4a14c4ebbc2f3e91e3aef3b329662291

I'm still working on it. I need to add group assignment feature.
If you think it could be useful also for others I can submit a PR in the coming days.

### PR DESCRIPTION
**Pull request by kenjis**
**Description**
Supersedes #579
Closes #566

Add command `shield:user` to manage users:
```
            shield:user create -n newusername -e newuser@example.com
```
```
            shield:user activate -n username
            shield:user activate -e user@example.com
```
```
            shield:user deactivate -n username
            shield:user deactivate -e user@example.com
```
```
            shield:user changename -n username --new-name newusername
            shield:user changename -e user@example.com --new-name newusername
```
```
            shield:user changeemail -n username --new-email newuseremail@example.com
            shield:user changeemail -e user@example.com --new-email newuseremail@example.com
```
```
            shield:user delete -i 123
            shield:user delete -n username
            shield:user delete -e user@example.com
```
```
            shield:user password -n username
            shield:user password -e user@example.com
```
```
            shield:user list
            shield:user list -n username -e user@example.com
```
```
            shield:user addgroup -n username -g mygroup
            shield:user addgroup -e user@example.com -g mygroup
```
```
            shield:user removegroup -n username -g mygroup
            shield:user removegroup -e user@example.com -g mygroup
```

TODO
- [x] remove `exit()`
- [x] fix if conditions
- [x] fix hard-coded table names
- [x] fix hard-coded validation rules

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [x] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: src/Commands/User.php [hunk 1]
```diff
@@ -0,0 +1,693 @@
...
+        $passwordConfirm = $this->prompt(
+            'Password confirmation',
+            null,
+            $this->validationRules['password']
+        );
+
+        if ($password !== $passwordConfirm) {
+            $this->write("The passwords don't match", 'red');
+
+            throw new RuntimeException("The passwords don't match");
```
Comment (datamweb) [inline src/Commands/User.php:288]: Not good. Do not remove the user from the process. Display new form to re-enter data.

File: src/Commands/User.php [hunk 2]
```diff
@@ -0,0 +1,693 @@
...
+     * @param string|null $newUsername User new name (optional)
+     */
+    private function changename(
+        ?string $username = null,
+        ?string $email = null,
+        ?string $newUsername = null
+    ): void {
+        $validation = Services::validation();
+        $validation->setRules([
+            'username' => 'required|is_unique[users.username]',
```
Comment (datamweb) [inline src/Commands/User.php:376]: Get table name from config file.

File: src/Commands/User.php [hunk 3]
```diff
@@ -0,0 +1,693 @@
...
+        if ($username !== null) {
+            $user = $userModel->where('username', $username)->first();
+        } elseif ($email !== null) {
+            $user = $userModel->where('secret', $email)->first();
+        }
+
+        if ($user === null) {
+            $this->write("User doesn't exist", 'red');
+
+            throw new RuntimeException("User doesn't exist");
```
Comment (datamweb) [inline src/Commands/User.php]: In all cases, I disagree with throw.

File: src/Controllers/RegisterController.php [hunk 4]
```diff
@@ -175,37 +165,10 @@ protected function getUserEntity(): User
      * @return array<string, array<string, array<string>|string>>
      * @phpstan-return array<string, array<string, string|list<string>>>
      */
-    protected function getValidationRules(): array
+    public function getValidationRules(): array
```
Comment (datamweb) [inline src/Controllers/RegisterController.php:168]: breaking change?

File: src/Validation/RegistrationValidationRules.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        $registrationUsernameRules = array_merge(
+            config('AuthSession')->usernameValidationRules,
+            [sprintf('is_unique[%s.username]', $this->tables['users'])]
+        );
+        $registrationEmailRules = array_merge(
+            config('AuthSession')->emailValidationRules,
+            [sprintf('is_unique[%s.secret]', $this->tables['identities'])]
+        );
+
+        return setting('Validation.registration') ?? [
```
Comment (datamweb) [inline src/Validation/RegistrationValidationRules.php:35]: Add `helper(['setting']);`

File: src/Commands/User.php [hunk 6]
```diff
@@ -0,0 +1,693 @@
...
+        '--new-email' => 'New email',
+        '-g'          => 'Group name',
+    ];
+
+    /**
+     * Validation rules for user fields
+     */
+    private array $validationRules = [
+        'username' => 'required|is_unique[users.username]',
+        'email'    => 'required|valid_email|is_unique[auth_identities.secret]',
```
[the remark this answers is not in the document]
Comment (kenjis) [reply src/Commands/User.php:126]: I forgot to remove it.
It should be removed.
See setValidationRules().

File: src/Commands/User.php (hunk 2 above)
Comment (kenjis) [reply src/Commands/User.php:372]: These code are not needed. I will remove.

File: src/Controllers/RegisterController.php (hunk 4 above)
Comment (kenjis) [reply src/Controllers/RegisterController.php]: Reverted.

File: src/Validation/RegistrationValidationRules.php (hunk 5 above)
Comment (kenjis) [reply src/Validation/RegistrationValidationRules.php:37]: Added.

File: src/Commands/User.php (hunk 3 above)
Comment (kenjis) [reply src/Commands/User.php]: The exception will be caught in `run()` method.
It is the replacement of `exit()`.

File: src/Commands/User.php [hunk 7]
```diff
@@ -0,0 +1,693 @@
...
+    public function run(array $params): int
+    {
+        $this->ensureInputOutput();
+        $this->setTables();
+        $this->setValidationRules();
+
+        $action = $params[0] ?? null;
+
+        if ($action === null || ! in_array($action, $this->validActions, true)) {
+            $this->write(
```
[the remark this answers is not in the document]
Comment (kenjis) [reply src/Commands/User.php:145]: This PR is too big. If you want to add more features, you should send another PR.

I am more concerned about the current features working properly than usability.

**Comment by kenjis** [comment]
@robertogerola @cijagani Try this, if you can.

**Review by cijagani — APPROVED** [review]

File: src/Commands/User.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (datamweb) [reply src/Commands/User.php:145]: > I am more concerned about the current features working properly than usability.

Ok I get it, I haven't exactly checked out this PR yet. I will try to continue with this in mind.

**Review by robertogerola — APPROVED** [review]

**Comment by kenjis** [comment]
Okay. The test code has been improved, so there should be no major bugs.
Testing this feature is welcome!

**Pull request merged**